store/tikv/region_request.go

```diff
@@ -263,22 +270,74 @@ func (s *RegionRequestSender) SendReqCtx(
 	}
 }
 
-func (s *RegionRequestSender) sendReqToRegion(bo *Backoffer, ctx *RPCContext, req *tikvrpc.Request, timeout time.Duration) (resp *tikvrpc.Response, retry bool, err error) {
-	if e := tikvrpc.SetContext(req, ctx.Meta, ctx.Peer); e != nil {
+// RPCCancellerCtxKey is context key attach rpc send cancelFunc collector to ctx.
+type RPCCancellerCtxKey struct{}
+
+// RPCCanceller is rpc send cancelFunc collector.
+type RPCCanceller struct {
+	sync.Mutex
+	allocID   int
+	cancels   map[int]func()
+	cancelled bool
+}
+
+// NewRPCanceller creates RPCCanceller with init state.
+func NewRPCanceller() *RPCCanceller {
+	return &RPCCanceller{cancels: make(map[int]func())}
+}
+
+// WithCancel generates new context with cancel func.
+func (h *RPCCanceller) WithCancel(ctx context.Context) (context.Context, func()) {
+	nctx, cancel := context.WithCancel(ctx)
+	h.Lock()
+	if h.cancelled {
+		h.Unlock()
+		cancel()
+		return nctx, func() {}
+	}
+	id := h.allocID
+	h.allocID++
+	h.cancels[id] = cancel
+	h.Unlock()
+	return nctx, func() {
+		cancel()
+		h.Lock()
+		delete(h.cancels, id)
```

Comment:
Does the `id` work here? because it is just a temporary variable at L:298

Reply:
the `id` will be caught by the cancel closure and help remove cancel from map, it just help avoid memory leak for after cancel be executed

Reply:
Why need to delete here? The `cancel` function can be repeated call, so the `delete` here is redundant?

Reply:
I think only a cancel function in copIterator is ok, becase copIteratorWorker can share the  cancel to stop work.